src/core/unit.c

```diff
@@ -5576,18 +5583,27 @@ void unit_unlink_state_files(Unit *u) {
         if (!u->id)
                 return;
 
-        if (!MANAGER_IS_SYSTEM(u->manager))
-                return;
-
         /* Undoes the effect of unit_export_state() */
 
         if (u->exported_invocation_id) {
-                p = strjoina("/run/systemd/units/invocation:", u->id);
+                if (MANAGER_IS_SYSTEM(u->manager))
+                        p = strjoina("/run/systemd/units/invocation:", u->id);
+                else {
+                        _cleanup_free_ char *user_path = NULL;
+                        if (xdg_user_runtime_dir(&user_path, "/systemd/units/invocation:") < 0) {
+                                log_unit_debug(u, "Failed to get user runtime units invocation path");
+                                return;
+                        }
+                        p = strjoina(user_path, u->id);
+                }
```

Comment:
given that this code is now a bit more complex and we need it twice. Can you make it a function of its own, returning a malloc()ed string? i.e.

unit_get_invocation_path() or so? (should use path_join() then internally)